drivers/display/display_framebuf.c

```diff
@@ -0,0 +1,251 @@
+ * should encounter no trouble. Avoid manipulating image data as byte arrays.
+ *
+ * Note that this implementation isn't particularly fast; it's aimed at
+ * simplicity. There's a lot of room for performance improvement.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <display.h>
+#include <display/framebuf.h>
+#include <string.h>
+
+#define GREY 0x00C0C0C0
+#define BLACK 0x00000000
+
+static int framebuf_blanking_on(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_blanking_off(const struct device *dev)
+{
+	return -ENOTSUP;
+}
+
+static void *framebuf_get_framebuffer(const struct device *dev)
+{
+	return NULL;
+}
+
+static int framebuf_set_brightness(const struct device *dev,
+				   const u8_t brightness)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_set_contrast(const struct device *dev,
+				 const u8_t contrast)
+{
+	return -ENOTSUP;
+}
+
+static int framebuf_set_pixel_format(const struct device *dev,
+				     const enum display_pixel_format format)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+
+	switch (format) {
+	case PIXEL_FORMAT_ARGB_8888:
+	case PIXEL_FORMAT_MONO01:
+	case PIXEL_FORMAT_MONO10:
+		data->format = format;
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int framebuf_set_orientation(const struct device *dev,
+				    const enum display_orientation orientation)
+{
+	switch (orientation) {
+	case DISPLAY_ORIENTATION_NORMAL:
+		return 0;
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static void framebuf_get_capabilities(const struct device *dev,
+				      struct display_capabilities *caps)
+{
+	struct framebuf_dev_data *data;
+
+	data = FRAMEBUF_DATA(dev);
+	caps->x_resolution = data->width;
+	caps->y_resolution = data->height;
+	caps->supported_pixel_formats = PIXEL_FORMAT_ARGB_8888 |
+					PIXEL_FORMAT_MONO01 |
+					PIXEL_FORMAT_MONO10;
+	caps->screen_info = 0;
+	caps->current_pixel_format = data->format;
+	caps->current_orientation = DISPLAY_ORIENTATION_NORMAL;
+}
+
+/*
+ * simple bit manipulation functions for display buffer read/write.
+ *
+ * we can't use the system-defined ones because they work on words,
+ * not bytes, which would introduce a dependency on endianness.
+ */
+
+static inline int test_bit(const void *buf, int bit)
+{
+	const u8_t *buf8 = buf;
+
+	return buf8[bit >> 3] & (1 << (7 - (bit & 7)));
+}
+
+static inline void set_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] |= (1 << (7 - (bit & 7)));
+}
+
+static inline void reset_bit(void *buf, int bit)
+{
+	u8_t *buf8 = buf;
+
+	buf8[bit >> 3] &= ~(1 << (7 - (bit & 7)));
+}
+
+static int framebuf_write(const struct device *dev, const u16_t x,
+			  const u16_t y,
+			  const struct display_buffer_descriptor *desc,
+			  const void *buf)
+{
+	struct framebuf_dev_data *data = FRAMEBUF_DATA(dev);
+	u32_t *dst = data->buffer;
+	const u32_t *src;
+	int row, col, bit;
+	u32_t set_color, reset_color;
+
+	dst += x;
+	dst += (y * data->pitch);
+
+	switch (data->format) {
```

Comment:
Please move the pixel format conversion to the upper layer graphics library (if needed) this driver should only be concerned with HW access.

Note that this API is only needed if direct frame buffer access is not possible.

Reply:
That would not be analogous with display_sdl.c which does the same kind of pixel conversion.

Reply:
The SDL based display is only intended for display simulation together with the native_posix target as such conversion functions are provided to prevent adaptation of the actual target code.

Unless this code is intended for simulation only, please move this code to the upper graphics implementation to prevent code duplication.

Reply:
So let me get this straight.  You want every client of the display code to be written to work on every possible display configuration, and that's to avoid code duplication?

Reply:
Yes, the actual drivers are intended to be thin layers between the display controllers and upper layer libraries (eg. CFB, LVGL). The intention is that clients interface with this upper layer libraries and not directly with the display drivers. 

In case there is common code that can be shared between the upper layer libraries this could be placed in a common support library.

Reply:
Ok, so you'd prefer me to remove framebuf_write() and framebuf_read(), because these APIs aren't necessary, and instead implement framebuf_get_framebuffer(), thus turning the framebuffer display driver into a glorified container for a few bits of metadata and a pointer?

And all of this is because our driver abstraction isn't supposed to abstract anything at all, that's all supposed to be handled in CFB, except that CFB doesn't handle it at all.

Reply:
You could leave them in place if you feel there is a need for them and also implement ``framebuf_get_framebuffer`` as such the upper layer library can decide which of the 2 variants it will use. But none of these functions should take care of color space conversion this should be handled by the upper layers.

The drivers should abstract how the display controller hardware is accessed (e.g. SPI, I2C, direct memory, ...) and provide a common API to upper layers. Currently we have 2 upper layer display libraries (with their limitations), CFB and LVGL.

Reply:
You have a vision.  That's great, but your vision is yet unrealized.  

I'll make a deal with you.  Once CFB is mature enough to be suitable to my task, I will happily remove the color-space conversion from my driver and restructure the display-based console around it.  Until then, let's put the code that works _today_ and solves a problem _today_ in the tree.

Reply:
There is a PR about support an other pixel format in CFB, please take a look at #16331

Reply:
@jfischer-phytec-iot Unfortunately there are other issues with CFB.  For example, it allocates and manipulates a back buffer, which is then flushed to the display in its entirety using a finalize function.  The insinuation that @vanwinkeljan makes is that I am responsible for fixing CFB to work the way that I need it to, and I'm just not sure that's a burden fairly placed on my shoulders.  As previously stated, if/when CFB suits my requirements, I'll use it. Until then I have a perfectly fine solution that is ready now, which is being blocked because its offends someone's sense of aesthetics.

Reply:
> The insinuation that @vanwinkeljan makes is that I am responsible for fixing CFB to work the way that I need it to, and I'm just not sure that's a burden fairly placed on my shoulders.

Hmm looks like we have a misunderstanding here, I asked if it was an option to use CFB (in a comment on drivers/console/display_console.c) from you replies and after a quick look it was clear it is not and I have no problem with that.

My question here was to move the color conversion from the driver to the upper layer, which is display_console in this case and not CFB.

Reply:
I've removed the color conversion from the framebuffer driver and eliminated `display_console`.